File: python/src/cosmoner/webhooks.py
"""Webhooks service namespace — endpoint management and delivery inspection."""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from ._config import ClientConfig, resolve_project_id
from ._transport import AsyncTransport, Transport

#: Every event type an endpoint can subscribe to.
WEBHOOK_EVENT_TYPES = (
    "email.sent",
    "email.delivered",
    "email.delivery_delayed",
    "email.bounced",
    "email.complained",
    "email.opened",
    "email.clicked",
    "email.rejected",
    "email.rendering_failed",
    "email.domain_verified",
    "email.sending_paused",
    "app.deployed",
    "app.failed",
    "domain.verified",
    "domain.expired",
    "server.running",
    "server.error",
    "member.invited",
    "member.joined",
)

#: Sentinel distinguishing "leave unchanged" from an explicit ``None``.
_UNSET = object()


def _create_payload(
    name: str, url: str, events: Sequence[str], description: str | None
) -> dict[str, Any]:
    """Validates create arguments and shapes them into the API request body."""
    if not name:
        raise ValueError("name is required")
    if not url:
        raise ValueError("url is required")
    if not events:
        raise ValueError("At least one event is required")

    payload: dict[str, Any] = {"name": name, "url": url, "events": list(events)}
    if description is not None:
        payload["description"] = description

    return payload


def _update_payload(
    name: str | None,
    url: str | None,
    events: Sequence[str] | None,
    description: Any,
    enabled: bool | None,
) -> dict[str, Any]:
    """Shapes update arguments, omitting anything the caller left untouched.

    ``description`` uses a sentinel rather than ``None`` because clearing a
    description means sending an explicit null.
    """
    if events is not None and not events:
        raise ValueError("At least one event is required")

    payload: dict[str, Any] = {}
    if name is not None:
        payload["name"] = name
    if url is not None:
        payload["url"] = url
    if events is not None:
        payload["events"] = list(events)
    if description is not _UNSET:
        payload["description"] = description
    if enabled is not None:
        payload["enabled"] = enabled

    return payload


def _delivery_params(
    status: str | None, event_type: str | None, limit: int | None, cursor: str | None
) -> dict[str, Any]:
    """Drops absent delivery filters so they are not sent as empty query values."""
    params: dict[str, Any] = {}
    if status is not None:
        params["status"] = status
    if event_type is not None:
        params["eventType"] = event_type
    if limit is not None:
        params["limit"] = limit
    if cursor is not None:
        params["cursor"] = cursor

    return params


def _require_endpoint_id(endpoint_id: str) -> None:
    """Rejects an empty endpoint id before it becomes a malformed route."""
    if not endpoint_id:
        raise ValueError("endpoint_id is required")


class WebhooksService:
    """Synchronous webhook endpoint and delivery operations for a project."""

    def __init__(self, transport: Transport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        return f"/v1/projects/{resolve_project_id(self._config, project_id)}/webhooks"

    def list(self, *, project_id: str | None = None) -> dict[str, Any]:
        """Lists every endpoint on the project, newest first."""
        result: dict[str, Any] = self._transport.request(
            "GET", self._base_path(project_id)
        )
        return result

    def get(self, endpoint_id: str, *, project_id: str | None = None) -> dict[str, Any]:
        """Fetches one endpoint together with its delivery counts."""
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = self._transport.request(
            "GET", f"{self._base_path(project_id)}/{endpoint_id}"
        )
        return result

    def create(
        self,
        name: str,
        url: str,
        events: Sequence[str],
        *,
        description: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Creates an endpoint and returns it with its signing secret.

        The secret is returned by this call alone — every later read gives only
        ``secretHint``, so persist it before discarding the response.
        """
        payload = _create_payload(name, url, events, description)

        result: dict[str, Any] = self._transport.request(
            "POST", self._base_path(project_id), json=payload
        )
        return result

    def update(
        self,
        endpoint_id: str,
        *,
        name: str | None = None,
        url: str | None = None,
        events: Sequence[str] | None = None,
        description: str | None = _UNSET,  # type: ignore[assignment]
        enabled: bool | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Updates an endpoint's configuration.

        Re-enabling a paused endpoint through ``enabled=True`` also clears its
        failure streak, so one more failure will not immediately re-pause it.
        Pass ``description=None`` to clear the description.
        """
        _require_endpoint_id(endpoint_id)
        payload = _update_payload(name, url, events, description, enabled)

        result: dict[str, Any] = self._transport.request(
            "PATCH", f"{self._base_path(project_id)}/{endpoint_id}", json=payload
        )
        return result

    def delete(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Deletes an endpoint and its delivery history."""
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = self._transport.request(
            "DELETE", f"{self._base_path(project_id)}/{endpoint_id}"
        )
        return result

    def resume(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Clears an auto-pause and re-queues the backlog held while it was down.

        An endpoint pauses itself after ten consecutive failed deliveries.
        """
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = self._transport.request(
            "POST", f"{self._base_path(project_id)}/{endpoint_id}/resume"
        )
        return result

    def rotate_secret(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Issues a new signing secret and returns it in full.

        The previous secret stops verifying immediately, so deploy the new one
        before rotating if the receiver cannot tolerate rejected deliveries.
        """
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = self._transport.request(
            "POST", f"{self._base_path(project_id)}/{endpoint_id}/rotate-secret"
        )
        return result

    def test(
        self,
        endpoint_id: str,
        *,
        event_type: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Sends a synthetic event and reports what the endpoint answered.

        Defaults to the endpoint's first subscribed event. Test sends never
        move the failure streak in either direction.
        """
        _require_endpoint_id(endpoint_id)
        payload = {"eventType": event_type} if event_type is not None else {}

        result: dict[str, Any] = self._transport.request(
            "POST", f"{self._base_path(project_id)}/{endpoint_id}/test", json=payload
        )
        return result

    def list_deliveries(
        self,
        endpoint_id: str,
        *,
        status: str | None = None,
        event_type: str | None = None,
        limit: int | None = None,
        cursor: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Lists deliveries for an endpoint, newest first, cursor-paginated."""
        _require_endpoint_id(endpoint_id)
        params = _delivery_params(status, event_type, limit, cursor)

        result: dict[str, Any] = self._transport.request(
            "GET",
            f"{self._base_path(project_id)}/{endpoint_id}/deliveries",
            params=params or None,
        )
        return result

    def replay_delivery(
        self, endpoint_id: str, delivery_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Queues a fresh delivery carrying the same payload.

        The endpoint must be enabled — replaying into a paused endpoint fails
        rather than silently queueing behind the backlog.
        """
        _require_endpoint_id(endpoint_id)
        if not delivery_id:
            raise ValueError("delivery_id is required")

        result: dict[str, Any] = self._transport.request(
            "POST",
            f"{self._base_path(project_id)}/{endpoint_id}/deliveries/{delivery_id}/replay",
        )
        return result


class AsyncWebhooksService:
    """Asynchronous counterpart to :class:`WebhooksService`."""

    def __init__(self, transport: AsyncTransport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        return f"/v1/projects/{resolve_project_id(self._config, project_id)}/webhooks"

    async def list(self, *, project_id: str | None = None) -> dict[str, Any]:
        """Lists every endpoint on the project, newest first."""
        result: dict[str, Any] = await self._transport.request(
            "GET", self._base_path(project_id)
        )
        return result

    async def get(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Fetches one endpoint together with its delivery counts."""
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = await self._transport.request(
            "GET", f"{self._base_path(project_id)}/{endpoint_id}"
        )
        return result

    async def create(
        self,
        name: str,
        url: str,
        events: Sequence[str],
        *,
        description: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Creates an endpoint and returns it with its signing secret.

        The secret is returned by this call alone — every later read gives only
        ``secretHint``, so persist it before discarding the response.
        """
        payload = _create_payload(name, url, events, description)

        result: dict[str, Any] = await self._transport.request(
            "POST", self._base_path(project_id), json=payload
        )
        return result

    async def update(
        self,
        endpoint_id: str,
        *,
        name: str | None = None,
        url: str | None = None,
        events: Sequence[str] | None = None,
        description: str | None = _UNSET,  # type: ignore[assignment]
        enabled: bool | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Updates an endpoint's configuration.

        Re-enabling a paused endpoint through ``enabled=True`` also clears its
        failure streak, so one more failure will not immediately re-pause it.
        Pass ``description=None`` to clear the description.
        """
        _require_endpoint_id(endpoint_id)
        payload = _update_payload(name, url, events, description, enabled)

        result: dict[str, Any] = await self._transport.request(
            "PATCH", f"{self._base_path(project_id)}/{endpoint_id}", json=payload
        )
        return result

    async def delete(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Deletes an endpoint and its delivery history."""
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = await self._transport.request(
            "DELETE", f"{self._base_path(project_id)}/{endpoint_id}"
        )
        return result

    async def resume(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Clears an auto-pause and re-queues the backlog held while it was down.

        An endpoint pauses itself after ten consecutive failed deliveries.
        """
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = await self._transport.request(
            "POST", f"{self._base_path(project_id)}/{endpoint_id}/resume"
        )
        return result

    async def rotate_secret(
        self, endpoint_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Issues a new signing secret and returns it in full.

        The previous secret stops verifying immediately, so deploy the new one
        before rotating if the receiver cannot tolerate rejected deliveries.
        """
        _require_endpoint_id(endpoint_id)

        result: dict[str, Any] = await self._transport.request(
            "POST", f"{self._base_path(project_id)}/{endpoint_id}/rotate-secret"
        )
        return result

    async def test(
        self,
        endpoint_id: str,
        *,
        event_type: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Sends a synthetic event and reports what the endpoint answered.

        Defaults to the endpoint's first subscribed event. Test sends never
        move the failure streak in either direction.
        """
        _require_endpoint_id(endpoint_id)
        payload = {"eventType": event_type} if event_type is not None else {}

        result: dict[str, Any] = await self._transport.request(
            "POST", f"{self._base_path(project_id)}/{endpoint_id}/test", json=payload
        )
        return result

    async def list_deliveries(
        self,
        endpoint_id: str,
        *,
        status: str | None = None,
        event_type: str | None = None,
        limit: int | None = None,
        cursor: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Lists deliveries for an endpoint, newest first, cursor-paginated."""
        _require_endpoint_id(endpoint_id)
        params = _delivery_params(status, event_type, limit, cursor)

        result: dict[str, Any] = await self._transport.request(
            "GET",
            f"{self._base_path(project_id)}/{endpoint_id}/deliveries",
            params=params or None,
        )
        return result

    async def replay_delivery(
        self, endpoint_id: str, delivery_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Queues a fresh delivery carrying the same payload.

        The endpoint must be enabled — replaying into a paused endpoint fails
        rather than silently queueing behind the backlog.
        """
        _require_endpoint_id(endpoint_id)
        if not delivery_id:
            raise ValueError("delivery_id is required")

        result: dict[str, Any] = await self._transport.request(
            "POST",
            f"{self._base_path(project_id)}/{endpoint_id}/deliveries/{delivery_id}/replay",
        )
        return result

File: javascript/src/services/webhooks.ts
/* eslint-disable require-await -- Every method here validates its arguments before
   reaching the transport. Keeping them `async` makes a bad argument reject the
   returned promise rather than throw synchronously, so one `.catch()` covers both
   client-side and server-side failures. */

/** Webhooks service namespace — endpoint management and delivery inspection. */

import { resolveProjectId, type ResolvedConfig } from "../config";
import type { Transport } from "../transport";
import {
  constructEvent,
  verifyWebhookSignature,
  type VerifyWebhookParams,
  type WebhookEvent,
} from "../webhooks/signature";

/** Every event type an endpoint can subscribe to. */
export const WEBHOOK_EVENT_TYPES = [
  "email.sent",
  "email.delivered",
  "email.delivery_delayed",
  "email.bounced",
  "email.complained",
  "email.opened",
  "email.clicked",
  "email.rejected",
  "email.rendering_failed",
  "email.domain_verified",
  "email.sending_paused",
  "app.deployed",
  "app.failed",
  "domain.verified",
  "domain.expired",
  "server.running",
  "server.error",
  "member.invited",
  "member.joined",
] as const;

export type WebhookEventType = (typeof WEBHOOK_EVENT_TYPES)[number];

/** Lifecycle state of a single delivery attempt chain. */
export type WebhookDeliveryStatus = "PENDING" | "SUCCEEDED" | "FAILED";

/** Why an endpoint stopped receiving events. */
export type WebhookDisabledReason = "MANUAL" | "CONSECUTIVE_FAILURES";

/** A configured webhook endpoint. The full secret is never returned here. */
export interface WebhookEndpoint {
  id: string;
  name: string;
  description: string | null;
  url: string;
  events: WebhookEventType[];
  enabled: boolean;
  /** Last four characters of the signing secret — enough to tell keys apart. */
  secretHint: string;
  consecutiveFailures: number;
  disabledAt: string | null;
  disabledReason: WebhookDisabledReason | null;
  lastSuccessAt: string | null;
  lastFailureAt: string | null;
  createdAt: string;
  updatedAt: string;
}

/** Delivery counts behind an endpoint's health badge. */
export interface WebhookEndpointStats {
  succeeded: number;
  failed: number;
  pending: number;
}

/** One attempt chain for one event against one endpoint. */
export interface WebhookDelivery {
  id: string;
  eventType: WebhookEventType;
  eventId: string;
  payload: unknown;
  status: WebhookDeliveryStatus;
  attempt: number;
  nextAttemptAt: string | null;
  responseStatus: number | null;
  responseBody: string | null;
  errorMessage: string | null;
  durationMs: number | null;
  deliveredAt: string | null;
  createdAt: string;
}

/** Options accepted by every method, for working across projects. */
export interface ProjectScopedParams {
  /** Overrides the client-level default project for this call. */
  projectId?: string;
}

/** Arguments accepted by `client.webhooks.create()`. */
export interface CreateWebhookEndpointParams extends ProjectScopedParams {
  name: string;
  /** Must be an HTTPS URL — the API rejects plaintext endpoints. */
  url: string;
  events: WebhookEventType[];
  description?: string;
}

/** Arguments accepted by `client.webhooks.update()`. */
export interface UpdateWebhookEndpointParams extends ProjectScopedParams {
  name?: string;
  url?: string;
  events?: WebhookEventType[];
  /** Pass null to clear the description. */
  description?: string | null;
  enabled?: boolean;
}

/** Filters accepted by `client.webhooks.listDeliveries()`. */
export interface ListDeliveriesParams extends ProjectScopedParams {
  status?: WebhookDeliveryStatus;
  eventType?: WebhookEventType;
  /** 1–100, defaults to 25 server-side. */
  limit?: number;
  /** `nextCursor` from the previous page. */
  cursor?: string;
}

/** An endpoint returned with its signing secret, shown only once. */
export interface WebhookEndpointWithSecret extends WebhookEndpoint {
  /**
   * The full signing secret. Returned only when the endpoint is created —
   * store it now, because no later call can retrieve it.
   */
  secret: string;
}

interface Envelope<T> {
  success: true;
  data: T;
}

export type ListWebhookEndpointsResponse = Envelope<WebhookEndpoint[]>;
export type GetWebhookEndpointResponse = Envelope<{
  endpoint: WebhookEndpoint;
  stats: WebhookEndpointStats;
}>;
export type CreateWebhookEndpointResponse = Envelope<WebhookEndpointWithSecret>;
export type UpdateWebhookEndpointResponse = Envelope<WebhookEndpoint>;
export type ResumeWebhookEndpointResponse = Envelope<WebhookEndpoint>;
export type RotateWebhookSecretResponse = Envelope<{ id: string; secret: string }>;
export type TestWebhookEndpointResponse = Envelope<{
  outcome: string;
  delivery: WebhookDelivery;
}>;
export type ListWebhookDeliveriesResponse = Envelope<{
  deliveries: WebhookDelivery[];
  nextCursor: string | null;
}>;
export type ReplayWebhookDeliveryResponse = Envelope<WebhookDelivery>;
export type DeleteWebhookEndpointResponse = Envelope<null>;

/** Webhook endpoint and delivery operations for a project. */
export class WebhooksService {
  constructor(
    private readonly transport: Transport,
    private readonly config: ResolvedConfig
  ) {}

  /** Builds the collection route for the resolved project. */
  private basePath(projectId?: string): string {
    return `/v1/projects/${resolveProjectId(this.config, projectId)}/webhooks`;
  }

  /** Lists every endpoint on the project, newest first. */
  async list(params: ProjectScopedParams = {}): Promise<ListWebhookEndpointsResponse> {
    return this.transport.request<ListWebhookEndpointsResponse>(
      "GET",
      this.basePath(params.projectId)
    );
  }

  /** Fetches one endpoint together with its delivery counts. */
  async get(
    endpointId: string,
    params: ProjectScopedParams = {}
  ): Promise<GetWebhookEndpointResponse> {
    if (!endpointId) throw new Error("endpointId is required");

    return this.transport.request<GetWebhookEndpointResponse>(
      "GET",
      `${this.basePath(params.projectId)}/${endpointId}`
    );
  }

  /**
   * Creates an endpoint and returns it with its signing secret.
   *
   * The secret is returned by this call alone — every later read gives only
   * `secretHint`, so persist it before discarding the response.
   */
  async create(params: CreateWebhookEndpointParams): Promise<CreateWebhookEndpointResponse> {
    if (!params.name) throw new Error("name is required");
    if (!params.url) throw new Error("url is required");
    if (!params.events?.length) throw new Error("At least one event is required");

    return this.transport.request<CreateWebhookEndpointResponse>(
      "POST",
      this.basePath(params.projectId),
      {
        body: {
          name: params.name,
          url: params.url,
          events: params.events,
          description: params.description,
        },
      }
    );
  }

  /**
   * Updates an endpoint's configuration.
   *
   * Re-enabling a paused endpoint through `enabled: true` also clears its
   * failure streak, so one more failure will not immediately re-pause it.
   */
  async update(
    endpointId: string,
    params: UpdateWebhookEndpointParams
  ): Promise<UpdateWebhookEndpointResponse> {
    if (!endpointId) throw new Error("endpointId is required");
    if (params.events && params.events.length === 0) {
      throw new Error("At least one event is required");
    }

    return this.transport.request<UpdateWebhookEndpointResponse>(
      "PATCH",
      `${this.basePath(params.projectId)}/${endpointId}`,
      {
        body: {
          name: params.name,
          url: params.url,
          events: params.events,
          description: params.description,
          enabled: params.enabled,
        },
      }
    );
  }

  /** Deletes an endpoint and its delivery history. */
  async delete(
    endpointId: string,
    params: ProjectScopedParams = {}
  ): Promise<DeleteWebhookEndpointResponse> {
    if (!endpointId) throw new Error("endpointId is required");

    return this.transport.request<DeleteWebhookEndpointResponse>(
      "DELETE",
      `${this.basePath(params.projectId)}/${endpointId}`
    );
  }

  /**
   * Clears an auto-pause and re-queues the backlog held while it was down.
   *
   * An endpoint pauses itself after ten consecutive failed deliveries.
   */
  async resume(
    endpointId: string,
    params: ProjectScopedParams = {}
  ): Promise<ResumeWebhookEndpointResponse> {
    if (!endpointId) throw new Error("endpointId is required");

    return this.transport.request<ResumeWebhookEndpointResponse>(
      "POST",
      `${this.basePath(params.projectId)}/${endpointId}/resume`
    );
  }

  /**
   * Issues a new signing secret and returns it in full.
   *
   * The previous secret stops verifying immediately, so deploy the new one
   * before rotating if the receiver cannot tolerate rejected deliveries.
   */
  async rotateSecret(
    endpointId: string,
    params: ProjectScopedParams = {}
  ): Promise<RotateWebhookSecretResponse> {
    if (!endpointId) throw new Error("endpointId is required");

    return this.transport.request<RotateWebhookSecretResponse>(
      "POST",
      `${this.basePath(params.projectId)}/${endpointId}/rotate-secret`
    );
  }

  /**
   * Sends a synthetic event and reports what the endpoint answered.
   *
   * Defaults to the endpoint's first subscribed event. Test sends never move
   * the failure streak in either direction.
   */
  async test(
    endpointId: string,
    params: ProjectScopedParams & { eventType?: WebhookEventType } = {}
  ): Promise<TestWebhookEndpointResponse> {
    if (!endpointId) throw new Error("endpointId is required");

    return this.transport.request<TestWebhookEndpointResponse>(
      "POST",
      `${this.basePath(params.projectId)}/${endpointId}/test`,
      { body: { eventType: params.eventType } }
    );
  }

  /** Lists deliveries for an endpoint, newest first, cursor-paginated. */
  async listDeliveries(
    endpointId: string,
    params: ListDeliveriesParams = {}
  ): Promise<ListWebhookDeliveriesResponse> {
    if (!endpointId) throw new Error("endpointId is required");

    return this.transport.request<ListWebhookDeliveriesResponse>(
      "GET",
      `${this.basePath(params.projectId)}/${endpointId}/deliveries`,
      {
        query: {
          status: params.status,
          eventType: params.eventType,
          limit: params.limit,
          cursor: params.cursor,
        },
      }
    );
  }

  /**
   * Queues a fresh delivery carrying the same payload.
   *
   * The endpoint must be enabled — replaying into a paused endpoint fails
   * rather than silently queueing behind the backlog.
   */
  async replayDelivery(
    endpointId: string,
    deliveryId: string,
    params: ProjectScopedParams = {}
  ): Promise<ReplayWebhookDeliveryResponse> {
    if (!endpointId) throw new Error("endpointId is required");
    if (!deliveryId) throw new Error("deliveryId is required");

    return this.transport.request<ReplayWebhookDeliveryResponse>(
      "POST",
      `${this.basePath(params.projectId)}/${endpointId}/deliveries/${deliveryId}/replay`
    );
  }

  /**
   * Reports whether a received request carries a valid signature.
   *
   * Needs the raw request body: a re-serialized object will not match what
   * was signed.
   */
  verify(params: VerifyWebhookParams): boolean {
    return verifyWebhookSignature(params);
  }

  /**
   * Verifies a received request and returns its parsed event envelope.
   *
   * Throws `WebhookSignatureError` when verification fails, so there is no
   * payload to act on unless the signature held.
   */
  constructEvent<T = unknown>(params: VerifyWebhookParams): WebhookEvent<T> {
    return constructEvent<T>(params);
  }
}
